Add invoke defaults for dialogs

keithamus opened the pull request:
#### dbcbbe87dd98b259cf8815a53c9bc7276892de6f
<pre>
Add invoke defaults for dialogs
<a href="https://bugs.webkit.org/show_bug.cgi?id=270602">https://bugs.webkit.org/show_bug.cgi?id=270602</a>

Reviewed by Anne van Kesteren.

This adds the `handleInvokeInternal` logic for HTMLDialogElement showmodal/close
a dialog.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-dialog-behavior.tentative-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-dialog-behavior.tentative-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-dialog-behavior.tentative-expected.txt.
* Source/WebCore/dom/Element.h:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::isValidInvokeAction):
(WebCore::HTMLDialogElement::handleInvokeInternal):
* Source/WebCore/html/HTMLDialogElement.h:
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::invokeAction const):

Canonical link: <a href="https://commits.webkit.org/278970@main">https://commits.webkit.org/278970@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/26243fa2e9282c7396e5eaeff79c1d9782144e31

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52102 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31430 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4518 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55375 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2814 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54406 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37819 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2522 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42398 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1797 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54196 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29057 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44953 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23469 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26327 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2230 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/984 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48234 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2373 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56972 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27221 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2391 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49798 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28457 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45072 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49059 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11404 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29362 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28198 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/WebCore/html/HTMLDialogElement.cpp:
```diff
@@ -128,6 +128,31 @@ void HTMLDialogElement::close(const String& result)
+    if (isPopoverShowing())
+        return false;
+
+    bool shouldClose = equalIgnoringASCIICase(action, "auto"_s) || 
```
should be `autoAtom()` instead of "auto"_s

lukewarlow commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1591,6 +1591,33 @@ PopoverState HTMLElement::popoverState() const
+        return false;
+    }
+
+    bool shouldShowPopover = equalIgnoringASCIICase(action, "auto"_s) || 
```
ditto. Should also declare atom strings for the other values.

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-popover-behavior.tentative-expected.txt:
```diff
@@ -1,20 +1,34 @@
 
 
-FAIL invoking (as auto) closed popover opens assert_true: expected true got false
+PASS invoking (as auto) closed popover opens
```
Are you able to now remove the additional expected file specific to ios?

lukewarlow commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1591,6 +1591,51 @@ PopoverState HTMLElement::popoverState() const
+        || equalIgnoringASCIICase(action, togglePopoverAtom())
+        || equalIgnoringASCIICase(action, hidePopoverAtom());
+
+    auto popoverState = popoverData();
```
nit: maybe call this popoverData so it's not confused with `popoverState()`

lukewarlow commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1591,6 +1591,51 @@ PopoverState HTMLElement::popoverState() const
+
+    auto popoverState = popoverData();
+
+    if (popoverState && popoverState->visibilityState() == PopoverVisibilityState::Hidden && shouldShowPopover) {
```
Nit this could be:

`if (!isPopoverShowing() && shouldShowPopover)`

lukewarlow commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1591,6 +1591,51 @@ PopoverState HTMLElement::popoverState() const
+        return true;
+    }
+
+    if (popoverState && popoverState->visibilityState() == PopoverVisibilityState::Showing && shouldHidePopover) {
```
nit this could be:

`if (isPopoverShowing() && shouldHIdePopover)`

lukewarlow commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1596,6 +1596,49 @@ PopoverState HTMLElement::popoverState() const
+        || equalIgnoringASCIICase(action, togglePopoverAtom())
+        || equalIgnoringASCIICase(action, hidePopoverAtom());
+
+    if (!isPopoverShowing() && shouldShowPopover)
```
Needs a { on the end

lukewarlow commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1596,6 +1596,49 @@ PopoverState HTMLElement::popoverState() const
+        return true;
+    }
+
+    if (isPopoverShowing() && shouldHidePopover)
```
ditto

lukewarlow commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -162,7 +162,7 @@ class HTMLElement : public StyledElement {
     void popoverAttributeChanged(const AtomString& value);
 
-    virtual void handleInvokeInternal(const AtomString&) { }
+    virtual bool handleInvokeInternal(const HTMLFormControlElement* invoker, const AtomString&);
```
The current change should be moved to the Element.h file (remember to return false in the default impl.
The below change should be made here.

```suggestion
   bool handleInvokeInternal(const HTMLFormControlElement* invoker, const AtomString&) override;
```

lukewarlow commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9718,6 +9718,8 @@ void Document::handlePopoverLightDismiss(const PointerEvent& event, Node& target
 
                     if (!invokerPopover) {
                         if (RefPtr button = dynamicDowncast<HTMLFormControlElement>(*htmlElement)) {
+                            if (RefPtr popover = button->invokeTargetElement(); popover && isShowingAutoPopover(*popover))
```
As invokeTargetElement now returns Element we need to dynamicDowncast it to HTMLElement as that's what isShowingAutoPopover() takes.

```suggestion
                            if (RefPtr popover = dynamicDowncast<HTMLElement>(button->invokeTargetElement()); popover && isShowingAutoPopover(*popover))
```

lukewarlow commented on Source/WebCore/html/HTMLFormControlElement.cpp:
```diff
@@ -444,6 +446,12 @@ InvokeAction HTMLFormControlElement::invokeAction() const
+        return InvokeAction::ShowModal;
+
+    if (equalLettersIgnoringASCIICase(action, closeLiteral))
+        return InvokeAction::ShowModal;
```
```suggestion
        return InvokeAction::Close;
```

annevk commented on Source/WebCore/html/HTMLDialogElement.cpp:
```diff
@@ -135,6 +135,36 @@ void HTMLDialogElement::close(const String& result)
+        return false;
+
+    if (isOpen()) {
+        const bool shouldClose = action == InvokeAction::Auto || action == InvokeAction::Close;
```
I think this should just be `auto`, maybe `bool`, but not `const`. Same below.

annevk reviewed (approved):
The iOS test issue needs to be addressed.

Modulo that and modulo the boolean nits this looks reasonable to me.

keithamus committed:
Add invoke defaults for dialogs
https://bugs.webkit.org/show_bug.cgi?id=270602

Reviewed by Anne van Kesteren.

This adds the `handleInvokeInternal` logic for HTMLDialogElement showmodal/close
a dialog.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-dialog-behavior.tentative-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-dialog-behavior.tentative-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-dialog-behavior.tentative-expected.txt.
* Source/WebCore/dom/Element.h:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::isValidInvokeAction):
(WebCore::HTMLDialogElement::handleInvokeInternal):
* Source/WebCore/html/HTMLDialogElement.h:
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::invokeAction const):

Canonical link: https://commits.webkit.org/278970@main
